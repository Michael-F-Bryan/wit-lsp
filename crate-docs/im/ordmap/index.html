<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="An ordered map."><title>im::ordmap - Rust</title><link rel="preload" as="font" type="font/woff2" crossorigin href="../../static.files/SourceSerif4-Regular-46f98efaafac5295.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../static.files/FiraSans-Regular-018c141bf0843ffd.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../static.files/FiraSans-Medium-8f9a781e4970d388.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../static.files/SourceCodePro-Regular-562dcc5011b6de7d.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../static.files/SourceCodePro-Semibold-d899c5a5c4aeb14a.ttf.woff2"><link rel="stylesheet" href="../../static.files/normalize-76eba96aa4d2e634.css"><link rel="stylesheet" href="../../static.files/rustdoc-5bc39a1768837dd0.css"><meta name="rustdoc-vars" data-root-path="../../" data-static-root-path="../../static.files/" data-current-crate="im" data-themes="" data-resource-suffix="" data-rustdoc-version="1.77.2 (25ef9e3d8 2024-04-09)" data-channel="1.77.2" data-search-js="search-dd67cee4cfa65049.js" data-settings-js="settings-4313503d2e1961c2.js" ><script src="../../static.files/storage-4c98445ec4002617.js"></script><script defer src="../sidebar-items.js"></script><script defer src="../../static.files/main-48f368f3872407c8.js"></script><noscript><link rel="stylesheet" href="../../static.files/noscript-04d5337699b92874.css"></noscript><link rel="alternate icon" type="image/png" href="../../static.files/favicon-16x16-8b506e7a72182f1c.png"><link rel="alternate icon" type="image/png" href="../../static.files/favicon-32x32-422f7d1d52889060.png"><link rel="icon" type="image/svg+xml" href="../../static.files/favicon-2c020d218678b618.svg"></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../../im/index.html">im</a><span class="version">15.1.0</span></h2></div><h2 class="location"><a href="#">Module ordmap</a></h2><div class="sidebar-elems"><section><ul class="block"><li><a href="#macros">Macros</a></li><li><a href="#structs">Structs</a></li><li><a href="#enums">Enums</a></li></ul></section><h2><a href="../index.html">In crate im</a></h2></div></nav><div class="sidebar-resizer"></div>
    <main><div class="width-limiter"><nav class="sub"><form class="search-form"><span></span><div id="sidebar-button" tabindex="-1"><a href="../../im/all.html" title="show sidebar"></a></div><input class="search-input" name="search" aria-label="Run search in the documentation" autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"><div id="help-button" tabindex="-1"><a href="../../help.html" title="help">?</a></div><div id="settings-menu" tabindex="-1"><a href="../../settings.html" title="settings"><img width="22" height="22" alt="Change settings" src="../../static.files/wheel-7b819b6101059cd0.svg"></a></div></form></nav><section id="main-content" class="content"><div class="main-heading"><h1>Module <a href="../index.html">im</a>::<wbr><a class="mod" href="#">ordmap</a><button id="copy-path" title="Copy item path to clipboard"><img src="../../static.files/clipboard-7571035ce49a181d.svg" width="19" height="18" alt="Copy item path"></button></h1><span class="out-of-band"><a class="src" href="../../src/im/ord/map.rs.html#5-2649">source</a> · <button id="toggle-all-docs" title="collapse all docs">[<span>&#x2212;</span>]</button></span></div><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>An ordered map.</p>
<p>An immutable ordered map implemented as a [B-tree] <a href="https://en.wikipedia.org/wiki/B-tree">1</a>.</p>
<p>Most operations on this type of map are O(log n). A
<a href="../hashmap/struct.HashMap.html"><code>HashMap</code></a> is usually a better choice for
performance, but the <code>OrdMap</code> has the advantage of only requiring
an <a href="https://doc.rust-lang.org/std/cmp/trait.Ord.html"><code>Ord</code></a> constraint on the key, and of being
ordered, so that keys always come out from lowest to highest,
where a <a href="../hashmap/struct.HashMap.html"><code>HashMap</code></a> has no guaranteed ordering.</p>
</div></details><h2 id="macros" class="section-header">Macros<a href="#macros" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="macro" href="macro.ordmap.html" title="macro im::ordmap::ordmap">ordmap</a></div><div class="desc docblock-short">Construct a map from a sequence of key/value pairs.</div></li></ul><h2 id="structs" class="section-header">Structs<a href="#structs" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="struct" href="struct.ConsumingIter.html" title="struct im::ordmap::ConsumingIter">ConsumingIter</a></div><div class="desc docblock-short">A consuming iterator over an ordered set.</div></li><li><div class="item-name"><a class="struct" href="struct.DiffIter.html" title="struct im::ordmap::DiffIter">DiffIter</a></div><div class="desc docblock-short">An iterator over the differences between two maps.</div></li><li><div class="item-name"><a class="struct" href="struct.Iter.html" title="struct im::ordmap::Iter">Iter</a></div><div class="desc docblock-short">An iterator over the key/value pairs of a map.</div></li><li><div class="item-name"><a class="struct" href="struct.Keys.html" title="struct im::ordmap::Keys">Keys</a></div><div class="desc docblock-short">An iterator ove the keys of a map.</div></li><li><div class="item-name"><a class="struct" href="struct.NodeDiffIter.html" title="struct im::ordmap::NodeDiffIter">NodeDiffIter</a></div><div class="desc docblock-short">An iterator over the differences between two ordered sets.</div></li><li><div class="item-name"><a class="struct" href="struct.OccupiedEntry.html" title="struct im::ordmap::OccupiedEntry">OccupiedEntry</a></div><div class="desc docblock-short">An entry for a mapping that already exists in the map.</div></li><li><div class="item-name"><a class="struct" href="struct.OrdMap.html" title="struct im::ordmap::OrdMap">OrdMap</a></div><div class="desc docblock-short">An ordered map.</div></li><li><div class="item-name"><a class="struct" href="struct.OrdMapPool.html" title="struct im::ordmap::OrdMapPool">OrdMapPool</a></div><div class="desc docblock-short">A memory pool for the appropriate node type.</div></li><li><div class="item-name"><a class="struct" href="struct.RangedIter.html" title="struct im::ordmap::RangedIter">RangedIter</a></div><div class="desc docblock-short">An iterator over an ordered set.</div></li><li><div class="item-name"><a class="struct" href="struct.VacantEntry.html" title="struct im::ordmap::VacantEntry">VacantEntry</a></div><div class="desc docblock-short">An entry for a mapping that does not already exist in the map.</div></li><li><div class="item-name"><a class="struct" href="struct.Values.html" title="struct im::ordmap::Values">Values</a></div><div class="desc docblock-short">An iterator over the values of a map.</div></li></ul><h2 id="enums" class="section-header">Enums<a href="#enums" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="enum" href="enum.DiffItem.html" title="enum im::ordmap::DiffItem">DiffItem</a></div><div class="desc docblock-short">A description of a difference between two ordered maps.</div></li><li><div class="item-name"><a class="enum" href="enum.Entry.html" title="enum im::ordmap::Entry">Entry</a></div><div class="desc docblock-short">A handle for a key and its associated value.</div></li><li><div class="item-name"><a class="enum" href="enum.NodeDiffItem.html" title="enum im::ordmap::NodeDiffItem">NodeDiffItem</a></div><div class="desc docblock-short">A description of a difference between two ordered sets.</div></li></ul></section></div></main></body></html>