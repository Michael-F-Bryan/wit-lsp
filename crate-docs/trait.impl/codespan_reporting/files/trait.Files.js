(function() {var implementors = {
"codespan_reporting":[],
"wit_compiler":[["impl&lt;'db&gt; <a class=\"trait\" href=\"codespan_reporting/files/trait.Files.html\" title=\"trait codespan_reporting::files::Files\">Files</a>&lt;'db&gt; for <a class=\"struct\" href=\"wit_compiler/queries/struct.WorkspaceFiles.html\" title=\"struct wit_compiler::queries::WorkspaceFiles\">WorkspaceFiles</a>&lt;'db&gt;"]]
};if (window.register_implementors) {window.register_implementors(implementors);} else {window.pending_implementors = implementors;}})()