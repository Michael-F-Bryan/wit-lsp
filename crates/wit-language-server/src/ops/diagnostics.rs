use codespan_reporting::diagnostic::{Label, LabelStyle, Severity};
use tower_lsp::lsp_types::{
    DiagnosticRelatedInformation, DiagnosticSeverity, DocumentDiagnosticReport,
    DocumentDiagnosticReportResult,
};
use wit_compiler::{
    diagnostics::{Diagnostics, Location},
    queries::{FilePath, SourceFile, Workspace},
};

use crate::utils;

pub fn file_diagnostics(
    db: &dyn wit_compiler::Db,
    ws: Workspace,
    file: SourceFile,
) -> DocumentDiagnosticReportResult {
    let path = file.path(db);
    let pkg = wit_compiler::queries::workspace_packages(db, ws)
        .into_iter()
        .find(|pkg| pkg.contains(db, path))
        .expect("unreachable");

    let diags = wit_compiler::queries::lower_package::accumulated::<Diagnostics>(db, ws, pkg);
    let items = diags
        .into_iter()
        .filter_map(|diag| lsp_diagnostic(db, ws, diag, file.path(db).raw_path(db)))
        .collect();

    DocumentDiagnosticReportResult::Report(DocumentDiagnosticReport::Full(
        tower_lsp::lsp_types::RelatedFullDocumentDiagnosticReport {
            full_document_diagnostic_report: tower_lsp::lsp_types::FullDocumentDiagnosticReport {
                items,
                ..Default::default()
            },
            ..Default::default()
        },
    ))
}

/// Convert a [`wit_compiler::diagnostics::Diagnostic`] to a
/// [`tower_lsp::lsp_types::Diagnostic`].
///
/// # Implementation Detail
///
/// This stashes a hash of the [`wit_compiler::diagnostics::Diagnostic`] inside
/// the resulting [`tower_lsp::lsp_types::Diagnostic`] so it can be retrieved
/// at a later date (e.g. when running a *Code Action* that can fix the issue).
/// The hash code is generated using [`utils::hash_diagnostic()`].
fn lsp_diagnostic(
    db: &dyn wit_compiler::Db,
    ws: Workspace,
    diag: wit_compiler::diagnostics::Diagnostic,
    uri: &str,
) -> Option<tower_lsp::lsp_types::Diagnostic> {
    let location = diag.location();

    if location.filename.raw_path(db) != uri {
        return None;
    }

    let hash = utils::hash_diagnostic(&diag);
    let diag = diag.as_diagnostic();
    let code = diag.code.as_deref()?;
    let msg = &diag.message;

    let secondaries: Vec<_> = diag
        .labels
        .iter()
        .filter(|label| label.style != LabelStyle::Primary)
        .collect();

    let related_information = secondaries
        .iter()
        .filter_map(|label| {
            let location = label_location(db, ws, label)?;
            Some(DiagnosticRelatedInformation {
                location: utils::location_to_lsp(db, location),
                message: label.message.clone(),
            })
        })
        .collect();

    Some(tower_lsp::lsp_types::Diagnostic {
        range: utils::ts_to_range(location.range),
        message: msg.to_string(),
        related_information: Some(related_information),
        code: Some(tower_lsp::lsp_types::NumberOrString::String(
            code.to_string(),
        )),
        severity: Some(lsp_severity(diag.severity)),
        data: Some(hash.into()),
        ..Default::default()
    })
}

fn lsp_severity(severity: Severity) -> DiagnosticSeverity {
    match severity {
        Severity::Bug | Severity::Error => DiagnosticSeverity::ERROR,
        Severity::Warning => DiagnosticSeverity::WARNING,
        Severity::Note => DiagnosticSeverity::INFORMATION,
        Severity::Help => DiagnosticSeverity::HINT,
    }
}

fn label_location(
    db: &dyn wit_compiler::Db,
    ws: Workspace,
    label: &Label<FilePath>,
) -> Option<Location> {
    let files = ws.files(db);
    let path = label.file_id;
    let line_numbers = wit_compiler::queries::calculate_line_numbers(db, files[&path]);

    let std::ops::Range { start, end } = label.range;
    let start_point = line_numbers.point(start).ok()?;
    let end_point = line_numbers.point(end).ok()?;
    let range = tree_sitter::Range {
        start_byte: start,
        end_byte: end,
        start_point,
        end_point,
    };

    Some(Location::new(path, range))
}
