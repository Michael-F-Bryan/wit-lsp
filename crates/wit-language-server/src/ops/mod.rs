mod fold;

pub use self::fold::folding_range;
