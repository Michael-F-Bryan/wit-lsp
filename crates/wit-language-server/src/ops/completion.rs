use tower_lsp::lsp_types::{CompletionItem, CompletionItemKind};
use wit_compiler::{
    access::ScopeIndex,
    ast::{self, AstNode, HasSource},
    queries::{metadata::HasIdent, SourceFile, Workspace},
    Db,
};

pub fn complete(
    db: &dyn Db,
    ws: Workspace,
    file: SourceFile,
    point: tree_sitter::Point,
) -> Vec<Completion> {
    let mut completions = Vec::new();

    // First, add all known keywords
    let keywords = wit_compiler::ast::KEYWORDS.iter().map(|kw| Completion {
        text: kw.to_string(),
        kind: CompletionKind::Keyword,
    });
    completions.extend(keywords);

    // Now, we do the hard job of figuring out which identifiers are in scope.
    // FIXME: This doesn't take imported items into account
    let path = file.path(db);
    let pkg = wit_compiler::queries::workspace_packages(db, ws)
        .into_iter()
        .find(|pkg| pkg.contains(db, path))
        .expect("unreachable");

    let items = wit_compiler::queries::package_items(db, pkg);

    if let Some(index) = items.enclosing_item(db, point) {
        let types = match index {
            ScopeIndex::World(index) => items.get_world(db, index).definitions(db),
            ScopeIndex::Interface(index) => items
                .get_interface(db, index)
                .items(db)
                .into_iter()
                .filter_map(|item| match item {
                    wit_compiler::queries::metadata::InterfaceItemMetadata::Func(_) => None,
                    wit_compiler::queries::metadata::InterfaceItemMetadata::Type(ty) => Some(ty),
                })
                .collect(),
        };

        let names = types
            .iter()
            .map(|t| t.ident(db).raw(db).to_string())
            .map(Completion::type_name);
        completions.extend(names);
    }

    let ast = wit_compiler::queries::parse(db, file);
    if let Some(ident) = ast.tree(db).ancestors(point).find_map(ast::Id::cast) {
        let src = file.contents(db);
        let ident = ident.utf8_text(src);
        // The user has started writing an identifier, so limit the completions
        // to whatever might match what they've written.
        completions.retain(|c| c.text.starts_with(ident));
    }

    // completions.sort_by(|left, right| left.text.cmp(&right.text));

    completions
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct Completion {
    pub text: String,
    pub kind: CompletionKind,
}

impl Completion {
    pub fn new(text: impl Into<String>, kind: CompletionKind) -> Self {
        Self {
            text: text.into(),
            kind,
        }
    }

    fn type_name(text: impl Into<String>) -> Self {
        Completion::new(text, CompletionKind::TypeName)
    }
}

impl From<Completion> for CompletionItem {
    fn from(value: Completion) -> Self {
        CompletionItem {
            label: value.text,
            kind: Some(value.kind.into()),
            ..Default::default()
        }
    }
}

#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
pub enum CompletionKind {
    TypeName,
    Interface,
    World,
    Keyword,
}

impl From<CompletionKind> for CompletionItemKind {
    fn from(value: CompletionKind) -> Self {
        match value {
            CompletionKind::TypeName => CompletionItemKind::STRUCT,
            CompletionKind::Interface => CompletionItemKind::INTERFACE,
            CompletionKind::World => CompletionItemKind::MODULE,
            CompletionKind::Keyword => CompletionItemKind::KEYWORD,
        }
    }
}

#[cfg(test)]
mod tests {
    use std::borrow::Cow;

    use im::OrdMap;
    use wit_compiler::{queries::FilePath, Compiler, Tree};

    use super::*;

    fn find_completion_location<'a>(
        tree: &Tree,
        src: &'a str,
    ) -> OrdMap<Cow<'a, str>, tree_sitter::Point> {
        tree.iter()
            .filter_map(wit_compiler::ast::Comment::cast)
            .map(|comment| (comment.text(src), comment.range().start_point))
            .collect()
    }

    #[test]
    fn completions_in_different_contexts() {
        let src = r#"
            interface foo {
                /* first */

                type x = /* second */ u32;
            }

            world bar {
                export /* third */;
            }
        "#;
        let db = Compiler::default();
        let path = FilePath::new(&db, "complete_builtins.wit".into());
        let file = SourceFile::new(&db, path, src.into());
        let ws = Workspace::new(&db, OrdMap::unit(path, file));
        let ast = wit_compiler::queries::parse(&db, file);
        let locations = find_completion_location(ast.tree(&db), src);

        let result: OrdMap<String, Vec<Completion>> = locations
            .into_iter()
            .map(|(k, p)| (k.to_string(), complete(&db, ws, file, p)))
            .collect();

        insta::assert_debug_snapshot!(result);
    }
}
