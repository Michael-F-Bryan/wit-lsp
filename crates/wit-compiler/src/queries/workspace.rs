use std::ops::Range;

use codespan_reporting::files::Error as CodespanError;
use im::{OrdMap, Vector};

use crate::{
    ast::{self, AstNode, HasSource},
    diagnostics::{Diagnostics, Location, MismatchedPackageDeclaration},
    Db, Text,
};

/// The path to a [`SourceFile`] in the [`Workspace`].
///
/// Ideally, this should only ever be passed around as an opaque identifier and
/// shown to the user. You shouldn't make any assumptions about its contents.
#[salsa::interned]
pub struct FilePath {
    #[return_ref]
    pub raw_path: Text,
}

/// A workspace keeps track of all files known to the compiler.
#[salsa::input]
#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub struct Workspace {
    pub files: OrdMap<FilePath, SourceFile>,
}

impl Workspace {
    /// Update a file's contents.
    pub fn update(self, db: &mut dyn Db, path: &str, text: impl Into<Text>) {
        let text = text.into();

        match self.lookup_by_path(db, path) {
            Some(file) => {
                file.set_contents(db).to(text);
            }
            None => {
                let mut files = self.files(db);
                let path = FilePath::new(db, path.into());
                let file = SourceFile::new(db, path, text);
                files.insert(path, file);
                self.set_files(db).to(files);
            }
        }
    }

    pub fn lookup_by_path(self, db: &dyn Db, path: &str) -> Option<SourceFile> {
        let files = self.files(db);

        files
            .iter()
            .find_map(|(p, f)| (p.raw_path(db) == path).then_some(*f))
    }

    /// Get a view of this [`Workspace`] as a
    /// [`codespan_reporting::files::Files`] database that can be used when
    /// rendering [`Diagnostics`].
    pub fn as_codespan_files(self, db: &dyn Db) -> WorkspaceFiles<'_> {
        WorkspaceFiles { db, ws: self }
    }
}

fn resolve_id(db: &dyn Db, files: &Vector<SourceFile>) -> Option<PackageId> {
    let mut id: Option<(PackageId, Location)> = None;

    for f in files.iter().copied() {
        let ast = crate::queries::parse(db, f);

        if let Some(node) = ast.source_file(db).package_opt() {
            if let Some(new_id) = package_id(db, node, f) {
                let location = Location::new(f.path(db), node.range());

                match &id {
                    Some((original_id, _)) if new_id == *original_id => {
                        // Same ID, so nothing to complain about
                    }
                    Some((original_id, original_definition)) => {
                        let diag = MismatchedPackageDeclaration {
                            original_definition: *original_definition,
                            original_id: *original_id,
                            second_id: new_id,
                            second_location: location,
                        };
                        Diagnostics::push(db, diag.into());
                    }
                    None => {
                        id = Some((new_id, location));
                    }
                }
            }
        }
    }

    id.map(|(id, _)| id)
}

#[salsa::tracked]
pub fn workspace_packages(db: &dyn Db, ws: Workspace) -> Vector<Package> {
    let files = ws.files(db);
    let mut folders: OrdMap<&str, Vector<SourceFile>> = OrdMap::new();

    for (path, f) in &files {
        // Note: We're kinda breaking FilePath's encapsulation by assuming
        // the path has some sort of internal structure (i.e. it contains
        // slash-separated segments), but this should be okay because in
        // practice we'll only ever be receiving URIs or fs paths, and both
        // satisfy that assumption.
        let path = path.raw_path(db);
        let dir = match path.rsplit_once(|c| c == '/' || c == '\\') {
            Some((dir, _)) => dir,
            None => ".",
        };

        let entries = folders.entry(dir).or_default();
        entries.push_back(*f);
        entries.sort();
    }

    folders
        .into_iter()
        .map(|(dir, files)| {
            let dir = FilePath::new(db, dir.into());
            let id = resolve_id(db, &files);
            Package::new(db, dir, id, files)
        })
        .collect()
}

fn package_id(db: &dyn Db, node: ast::PackageDecl<'_>, file: SourceFile) -> Option<PackageId> {
    let node = node.fully_qualified_package_name()?;
    let src = file.contents(db);

    let pkg = node.package()?.utf8_text(src);
    let pkg = Vector::unit(Text::from(pkg));

    let path = node
        .path()?
        .iter_identifiers()
        .map(|ident| Text::from(ident.utf8_text(src)))
        .collect();

    let version = node.version_opt().map(|v| v.utf8_text(src)).map(Text::from);

    Some(PackageId::new(db, pkg, path, version))
}

/// A group of files in a [`Workspace`].
#[salsa::tracked]
pub struct Package {
    pub dir: FilePath,
    pub id: Option<PackageId>,
    pub files: Vector<SourceFile>,
}

#[salsa::interned]
pub struct PackageId {
    pub package: Vector<Text>,
    pub path: Vector<Text>,
    pub version: Option<Text>,
}

/// A file attached to a [`Workspace`].
#[salsa::input]
pub struct SourceFile {
    pub path: FilePath,
    #[return_ref]
    pub contents: Text,
}

/// An implementation of [`codespan_reporting::files::Files`] which will query
/// the [`Db`] as required.
#[derive(Copy, Clone)]
pub struct WorkspaceFiles<'db> {
    ws: Workspace,
    db: &'db dyn Db,
}

impl<'db> WorkspaceFiles<'db> {
    fn file(&self, path: FilePath) -> Result<SourceFile, CodespanError> {
        let files = self.ws.files(self.db);
        files.get(&path).copied().ok_or(CodespanError::FileMissing)
    }
}

impl<'db> codespan_reporting::files::Files<'db> for WorkspaceFiles<'db> {
    type FileId = FilePath;
    type Name = Text;
    type Source = &'db str;

    fn name(&'db self, id: FilePath) -> Result<Text, CodespanError> {
        let file = self.file(id)?;
        Ok(file.path(self.db).raw_path(self.db).clone())
    }

    fn source(&'db self, id: FilePath) -> Result<&'db str, CodespanError> {
        let file = self.file(id)?;
        Ok(file.contents(self.db))
    }

    fn line_index(&'db self, path: FilePath, byte_index: usize) -> Result<usize, CodespanError> {
        let file = self.file(path)?;
        crate::queries::calculate_line_numbers(self.db, file).line_index(byte_index)
    }

    fn line_range(
        &'db self,
        path: FilePath,
        line_index: usize,
    ) -> Result<Range<usize>, CodespanError> {
        let file = self.file(path)?;
        crate::queries::calculate_line_numbers(self.db, file).line_range(line_index)
    }
}

#[cfg(test)]
mod tests {
    use salsa::DebugWithDb;

    use crate::Compiler;

    use super::*;

    macro_rules! workspace {
        ($db:ident, { $( $path:expr => $contents:expr),* $(,)? }) => {{
            let mut files = OrdMap::new();
            $(
                let path = FilePath::new(&$db, $path.into());
                let f = SourceFile::new(&$db, path, $contents.into());
                files.insert(path, f);
            )*
            Workspace::new(&$db, files)
        }};
    }

    #[test]
    fn determine_packages() {
        let db = Compiler::default();

        let ws = workspace!(db, {
            "root.wit" => "",
            "simple/unnamed.wit" => "",
            "named/package.wit" => "package named:package;",
            "named/with/version.wit" => "package named:with/version@1.2.3;",
            "merge/empty.wit" => "",
            "merge/package-decl.wit" => "package merge:some-package",
        });

        let packages = workspace_packages(&db, ws);

        let diags = workspace_packages::accumulated::<Diagnostics>(&db, ws);
        assert!(diags.is_empty(), "{diags:?}");

        let debug: Vec<_> = packages.iter().map(|pkg| pkg.debug_all(&db)).collect();
        insta::assert_debug_snapshot!(debug);
    }

    #[test]
    fn conflicting_package_declarations() {
        let db = Compiler::default();

        let ws = workspace!(db, {
            "first.wit" => "package conflict:first",
            "second.wit" => "package conflict:second",
        });

        let diags = workspace_packages::accumulated::<Diagnostics>(&db, ws);
        assert_eq!(diags.len(), 1);
        insta::assert_debug_snapshot!(diags);
    }
}
