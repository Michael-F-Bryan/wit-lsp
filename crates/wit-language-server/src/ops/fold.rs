use im::Vector;
use tower_lsp::lsp_types::{FoldingRange, FoldingRangeKind};
use tree_sitter::{Node, QueryCursor};
use wit_compiler::queries::Ast;

use crate::Db;

// FIXME: The folded ranges aren't 100% correct here.
//
// They'll often include the trailing "}" even though our query should match the
// items between the "{" and "}". The param_list one is also a bit suspect
// because the folded region won't include the name of the first parameter.
const QUERY: &str = r"
[
    (block_comment)
    (slash_comment)
    (doc_comment)
    (world_item items: (world_items))
    (interface_item items: (interface_items))
    (param_list params: (named_type))
    (doc_comment)
    (resource_item methods: (resource_method))
    (variant_item cases: (variant_case))
    (record_item fields: (record_field))
    (flags_item cases: (flags_case))
    (enum_item cases: (enum_case))
] @normal

[
    (block_comment)
    (slash_comment)
] @comments

[
    (top_level_use_item)
] @imports
";

/// Implement [range folding][spec].
///
/// [spec]: https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_foldingRange
#[salsa::tracked]
#[tracing::instrument(level = "debug", skip_all)]
pub fn folding_range(db: &dyn Db, ast: Ast) -> Vector<FoldingRange> {
    let db = db.as_wit();
    let root = ast.root_node(db);
    let src = ast.src(db);

    folding_range_impl(root, src)
}

fn folding_range_impl(root: Node<'_>, src: &str) -> Vector<FoldingRange> {
    let mut cursor = QueryCursor::new();
    let language = tree_sitter_wit::language();
    let query = tree_sitter::Query::new(&language, QUERY).expect("The query was invalid");

    let capture_names = query.capture_names();

    cursor
        .matches(&query, root, src.as_bytes())
        .flat_map(|m| {
            let kind = match capture_names[m.pattern_index] {
                "comments" => Some(FoldingRangeKind::Comment),
                "imports" => Some(FoldingRangeKind::Imports),
                "normal" => None,
                other => unreachable!(
                    "Unknown query capture name, {other:?}, expected one of {capture_names:?}"
                ),
            };
            m.captures
                .iter()
                .map(move |cap| node_range(cap.node, kind.clone()))
        })
        .collect()
}

fn node_range(node: Node<'_>, kind: Option<FoldingRangeKind>) -> FoldingRange {
    let tree_sitter::Range {
        start_point,
        end_point,
        ..
    } = node.range();

    FoldingRange {
        start_line: start_point.row as u32,
        start_character: Some(start_point.column as u32),
        end_line: end_point.row as u32,
        end_character: Some(end_point.column as u32),
        kind,
        collapsed_text: None,
    }
}

#[cfg(test)]
mod tests {
    use wit_compiler::Tree;

    use super::*;

    #[test]
    fn fold_all_resources() {
        let src = include_str!("../../../../integration-tests/compile-pass/all-resources.wit");
        let tree = Tree::parse(src);

        let got = folding_range_impl(tree.root_node(), src);

        insta::assert_yaml_snapshot!(&got);
    }
}
