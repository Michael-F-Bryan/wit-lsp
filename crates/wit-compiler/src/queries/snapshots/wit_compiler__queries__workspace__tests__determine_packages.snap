---
source: crates/wit-compiler/src/queries/workspace.rs
expression: debug
---
[
    Package {
        [salsa id]: 0,
        dir: FilePath {
            [salsa id]: 6,
            raw_path: ".",
        },
        id: None,
        files: [
            SourceFile(
                Id {
                    value: 1,
                },
            ),
        ],
    },
    Package {
        [salsa id]: 1,
        dir: FilePath {
            [salsa id]: 7,
            raw_path: "merge",
        },
        id: Some(
            PackageId {
                [salsa id]: 0,
                raw: "merge:some-package",
            },
        ),
        files: [
            SourceFile(
                Id {
                    value: 5,
                },
            ),
            SourceFile(
                Id {
                    value: 6,
                },
            ),
        ],
    },
    Package {
        [salsa id]: 2,
        dir: FilePath {
            [salsa id]: 8,
            raw_path: "named",
        },
        id: Some(
            PackageId {
                [salsa id]: 1,
                raw: "some-namespace:package",
            },
        ),
        files: [
            SourceFile(
                Id {
                    value: 3,
                },
            ),
        ],
    },
    Package {
        [salsa id]: 3,
        dir: FilePath {
            [salsa id]: 9,
            raw_path: "named/with",
        },
        id: Some(
            PackageId {
                [salsa id]: 2,
                raw: "another-namespace:package@1.2.3",
            },
        ),
        files: [
            SourceFile(
                Id {
                    value: 4,
                },
            ),
        ],
    },
    Package {
        [salsa id]: 4,
        dir: FilePath {
            [salsa id]: 10,
            raw_path: "simple",
        },
        id: None,
        files: [
            SourceFile(
                Id {
                    value: 2,
                },
            ),
        ],
    },
]
